### ISSUE TITLE
[py2py3] Remove/replace dependency on nats-client library

### ISSUE THREAD

**Issue by mapellidario** [dmwm/WMCore#10074]
**Impact of the new feature**

WMCore in general, and perhaps other services depending on it.

**Is your feature request related to a problem? Please describe.**

Some packages built from dmwm/WMCore depend on `nats-client` ( [pypi](https://pypi.org/project/nats-client/), [specfile](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/py2-nats.spec) ), which only supports python2. 

List of dmwm/WMCore packages depending on `nats-client`

* [WMagent](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/wmagent.spec), imported by [cmsmonitoring](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/cmsmonitoring.spec).
* [ReqMgr2](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/reqmgr2.spec), imported by [cmsmonitoring](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/cmsmonitoring.spec).
* [ReqMgr2MS](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/reqmgr2ms.spec), imported by [cmsmonitoring](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/cmsmonitoring.spec).
* [Workqueue](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/workqueue.spec), imported by [cmsmonitoring](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/cmsmonitoring.spec).
* [Reqmon](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/reqmon.spec), imported by [cmsmonitoring](https://github.com/cms-sw/cmsdist/blob/comp_gcc630/cmsmonitoring.spec).

**Describe the solution you'd like**

There is a version of `nats` compatible with python 3 and it is available at

* [github.com - nats.py](https://github.com/nats-io/nats.py)
* [pypi - asyncio-nats-client](https://pypi.org/project/asyncio-nats-client/)

We would then have to choose which libirary to load at runtime, depending on the python interpreter version.

**Describe alternatives you've considered**

We could drop the dependency or find and alternative (I have not looked into it yet)

**Comment by vkuznet** [comment]
Dario,
CMSMonitoring supports both python2 and python3 version of NATS, please see here:
https://github.com/dmwm/CMSMonitoring/blob/master/src/python/CMSMonitoring/NATS.py#L38-L48
and https://github.com/dmwm/CMSMonitoring/blob/master/src/python/CMSMonitoring/nats3.py
Therefore, it is really a question how DMWM loads the nats client. Since so far DMWM is python2 based the nats python2 client is used.  I suggest that you provide more details how you envision the transition. If you will provide py2-nats and py3-nats specs then DMWM may load appropriate one based on python version.

**Comment by mapellidario** [comment]
This is great news! 

I admit that I did not check CMSMonitoring code before opening this issue. I simply noticed that WMCore was relying on py2-nats through CMSMonitoring just by looking at the specfiles and then opened the issue. Thanks for easing my research!
I have not looked into at how WMCore imports CMSMonitoring yet, but your suggestion of providing a `py3-nats` specfile and then delegating the choice of to CMSMonitoring seems reasonable. 
I will keep this issue open as I reminder that I will have to look into this when we have a py3 environment ready.

Thanks again!

**Comment by amaltaro** [comment]
Without reading the whole discussion (apologies for that).
If WMAgent depends on CMSMonitoring library, and CMSMonitoring is available either in python2 and python3, I'd assume that CMSMonitoring would bring the correct dependencies according to the python stack we are running on.

Having said that, we should not even list NATS as a dependency for WMAgent, since that is an indirect dependency coming from "external" libraries. Unless we also use NATS in other modes than from CMSMonitoring, in that case we do have to keep it in the requirements.txt file.

**Comment by amaltaro** [comment]
@mapellidario as we discussed in the past week, I'm setting this issue free such that others can pick it, if they want.

**Comment by goughes** [comment]
I reviewed everyone's comments here and took a look at the WMCore code to see if we really have this nats dependency. I cannot find any reference to nats directly or via the CMSMonitoring package. The only import from from CMSMonitoring in WMCore is `StompAMQ`, which is used by the five services listed in Dario's description above.

I think this nats dependency will not impact us unless we decide to use nats later, which (I believe) would require a py3-nats spec file to be created from the PyPI package `asyncio-nats-client` and the `Requires:` of https://github.com/cms-sw/cmsdist/blob/comp_gcc630/py3-cmsmonitoring.spec to be updated.

Let me know your thoughts @amaltaro, @vkuznet.

**Comment by vkuznet** [comment]
Erik, as far as I'm aware, only DBS uses NATS, and also Stefano asked me about publishing data through NATS (but I don't know status of this development). Therefore, pure WMCore will not be affected. And, as you correctly pointed out we'll need py3-nats if it will be required.

The usage of cms-monitoring is independent from NATS and does not require NATS to be installed.

**Comment by amaltaro** [comment]
Cool! It looks like we have a consensus in the sense that there is nothing to be done here, and that WMCore does not have any direct/indirect dependency on `nats-client`.

I had a quick look at the spec files and only the `cmsmonitoring.spec` (python2 version) depends on nats-client. The equivalent python3 cmsmonitoring spec file `py3-cmsmonitoring.spec` no longer has that dependency. During some time, our wmagentpy3 spec had a wrong dependency on `cmsmonitoring` instead of `py3-cmsmonitoring`, maybe that's where the confusion came from, sorry about that!

In short, our current python3 stack is safe and there is nothing to be modified. Thanks to everyone that contributed on this issue! Closing it as nothing to be done!
